editor/src/viewport_tools/tools/crop.rs

```diff
@@ -1,30 +1,334 @@
+use crate::consts::SELECTION_TOLERANCE;
+use crate::document::DocumentMessageHandler;
+use crate::frontend::utility_types::MouseCursorIcon;
+use crate::input::keyboard::Key;
+use crate::input::InputPreprocessorMessageHandler;
 use crate::layout::widgets::PropertyHolder;
 use crate::message_prelude::*;
-use crate::viewport_tools::tool::ToolActionHandlerData;
+use crate::viewport_tools::snapping::SnapHandler;
+use crate::viewport_tools::tool::{DocumentToolData, Fsm, ToolActionHandlerData};
 
+use graphene::intersection::Quad;
+
+use super::shared::transformation_cage::*;
+
+use glam::{DVec2, Vec2Swizzles};
 use serde::{Deserialize, Serialize};
 
 #[derive(Default)]
-pub struct Crop;
+pub struct Crop {
+	fsm_state: CropToolFsmState,
+	data: CropToolData,
+}
 
 #[remain::sorted]
 #[impl_message(Message, ToolMessage, Crop)]
 #[derive(PartialEq, Clone, Debug, Hash, Serialize, Deserialize)]
 pub enum CropMessage {
 	// Standard messages
-	// #[remain::unsorted]
-	// Abort,
+	#[remain::unsorted]
+	Abort,
+	#[remain::unsorted]
+	DocumentIsDirty,
 
 	// Tool-specific messages
-	MouseMove,
+	MouseDown,
+	MouseMove {
+		axis_align: Key,
+		centre: Key,
+	},
+	MouseUp,
```

Comment:
We should probably use the more general Pointer move here, the usage kinda differs throughout he codebase buf iirc we agreed that it should be PointerMove etc

Reply:
I have cleaned up in the rest of the codebase.

Reply:
Awesome!